dev-tools/omdb/src/bin/omdb/mgs/dashboard.rs

```diff
@@ -0,0 +1,1115 @@
+                gap.push(None);
+            }
+
+            let graph = self.graphs.get_mut(graph).unwrap();
+
+            for _ in 0..length {
+                graph.data(&gap[0..sids.len()]);
+            }
+        }
+    }
+
+    fn values(&mut self, values: &SensorValues) {
+        for (graph, sids) in &self.sids {
+            let mut data = vec![];
+
+            for sid in sids {
+                if let Some(value) = values.values.get(&sid) {
+                    data.push(*value);
+                } else {
+                    data.push(None);
+                }
+            }
+
+            let graph = self.graphs.get_mut(graph).unwrap();
+            graph.data(data.as_slice());
+        }
+
+        self.time = values.time;
+    }
+}
+
+fn run_dashboard<B: Backend>(
+    terminal: &mut Terminal<B>,
+    dashboard: &mut Dashboard,
+    force_update: bool,
+) -> Result<bool> {
+    let update = if crossterm::event::poll(Duration::from_secs(0))? {
+        if let Event::Key(key) = event::read()? {
+            match key.code {
+                KeyCode::Char('q') => return Ok(true),
+                KeyCode::Char('+') => dashboard.zoom_in(),
+                KeyCode::Char('-') => dashboard.zoom_out(),
+                KeyCode::Char('<') => dashboard.time_left(),
+                KeyCode::Char('>') => dashboard.time_right(),
+                KeyCode::Char('!') => dashboard.flip(),
+                KeyCode::Char('l') => {
+                    //
+                    // ^L -- form feed -- is historically used to clear and
+                    // redraw the screen.  And, notably, it is what dtach(1)
+                    // will send when attaching to a dashboard.  If we
+                    // see ^L, clear the terminal to force a total redraw.
+                    //
+                    if key.modifiers == KeyModifiers::CONTROL {
+                        terminal.clear()?;
+                    }
+                }
+                KeyCode::Up => dashboard.up(),
+                KeyCode::Down => dashboard.down(),
+                KeyCode::Right => dashboard.right(),
+                KeyCode::Left => dashboard.left(),
+                KeyCode::Esc => dashboard.esc(),
+                KeyCode::Tab => dashboard.tab(),
+                _ => {}
+            }
+        }
+        true
+    } else {
+        force_update
+    };
+
+    if update {
+        dashboard.update_data();
+        terminal.draw(|f| draw(f, dashboard))?;
+    }
+
+    Ok(false)
+}
+
+fn secs() -> Result<u64> {
+    let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
+    Ok(now.as_secs())
+}
+
+///
+/// Runs `omdb mgs dashboard`
+///
+pub(crate) async fn cmd_mgs_dashboard(
+    omdb: &crate::Omdb,
+    log: &slog::Logger,
+    mgs_args: &crate::mgs::MgsArgs,
+    args: &DashboardArgs,
+) -> Result<(), anyhow::Error> {
+    let mut input = if let Some(ref input) = args.sensors_args.input {
+        let file = File::open(input)?;
```

Comment:
Probably want something like

```suggestion
        let file = File::open(input).with_context(|| format!("failed to open {input}"))?;
```

so the error message includes the path.

`.context()` / `.with_context()` _might_ be useful on other `?`s in this function, but I'm less sure about them (and also suspect it's a lot less expected for any of them to fail in practice?).

Reply:
Yeah, that's a little nicer. As it stood, it didn't leave *too* much to the imagination:

```console
$ omdb mgs dashboard -i nonono
Error: No such file or directory (os error 2)
```

But being more explicit is better -- thank you!